server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdater.java

```diff
@@ -264,34 +273,85 @@ private void deleteOrUpdateParameters(DbSession dbSession, RuleUpdate update, Ru
+          .toList();
       } else {
         // Delete linked active rule params
-        for (ActiveRuleDto activeRuleDto : activeRules.get(customRule)) {
-          for (ActiveRuleParamDto activeRuleParamDto : activeRuleParams.get(activeRuleDto)) {
-            if (activeRuleParamDto.getKey().equals(key)) {
-              dbClient.activeRuleDao().deleteParam(dbSession, activeRuleDto, activeRuleParamDto);
-            }
-          }
-        }
+        from(activeRuleParamsByActiveRule.values())
+          .filter(new DeleteActiveRuleParams(dbSession, dbClient, key))
+          .toList();
       }
       paramKeys.add(key);
     }
   }
 
+  private static class ActiveRuleParamToActiveRule implements Function<ActiveRuleParamDto, ActiveRuleDto> {
+    private final Map<Integer, ActiveRuleDto> activeRuleById;
+
+    private ActiveRuleParamToActiveRule(Map<Integer, ActiveRuleDto> activeRuleById) {
+      this.activeRuleById = activeRuleById;
+    }
+
+    @Override
+    public ActiveRuleDto apply(@Nonnull ActiveRuleParamDto input) {
+      return activeRuleById.get(input.getActiveRuleId());
+    }
+  }
+
+  private static class UpdateOrInsertActiveRuleParams implements Predicate<ActiveRuleDto> {
+    private final DbSession dbSession;
+    private final DbClient dbClient;
+    private final RuleParamDto ruleParamDto;
+    private final Multimap<ActiveRuleDto, ActiveRuleParamDto> activeRuleParams;
+
+    private UpdateOrInsertActiveRuleParams(DbSession dbSession, DbClient dbClient, RuleParamDto ruleParamDto, Multimap<ActiveRuleDto, ActiveRuleParamDto> activeRuleParams) {
+      this.dbSession = dbSession;
+      this.dbClient = dbClient;
+      this.ruleParamDto = ruleParamDto;
+      this.activeRuleParams = activeRuleParams;
+    }
+
+    @Override
+    public boolean apply(@Nonnull ActiveRuleDto activeRuleDto) {
+      Map<String, ActiveRuleParamDto> activeRuleParamByKey = from(activeRuleParams.get(activeRuleDto))
+        .uniqueIndex(ActiveRuleDtoParamToKey.INSTANCE);
+      ActiveRuleParamDto activeRuleParamDto = activeRuleParamByKey.get(ruleParamDto.getName());
+      if (activeRuleParamDto != null) {
+        dbClient.activeRuleDao().updateParam(dbSession, activeRuleDto, activeRuleParamDto.setValue(ruleParamDto.getDefaultValue()));
+      } else {
+        dbClient.activeRuleDao().insertParam(dbSession, activeRuleDto, ActiveRuleParamDto.createFor(ruleParamDto).setValue(ruleParamDto.getDefaultValue()));
+      }
+      return true;
+    }
+  }
+
+  private static class DeleteActiveRuleParams implements Predicate<ActiveRuleParamDto> {
+    private final DbSession dbSession;
+    private final DbClient dbClient;
+    private final String key;
+
+    public DeleteActiveRuleParams(DbSession dbSession, DbClient dbClient, String key) {
+      this.dbSession = dbSession;
+      this.dbClient = dbClient;
+      this.key = key;
+    }
+
+    @Override
+    public boolean apply(@Nonnull ActiveRuleParamDto activeRuleParamDto) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)